packages/dds/sequence/src/plan-for-updating-interval-api.md

```diff
@@ -0,0 +1,87 @@
+# Refactoring/Deprecating Interval-related API
+
+## Refactoring
+
+
+1. Keep the IntervalCollection related contents in the current file, it includes:
+
+- `IntervalCollection`: Collection of intervals that supports addition, modification, removal, and efficient spatial querying.
+
+- `LocalIntervalCollection`
+
+- `IIntervalCollectionEvent`
+
+- `IntervalCollectionIterator`
+
+- All corresponding interfaces and helper functions for `IntervalCollection`
+
+2.  Create a new folder for Intervals-related components
+
+- `intervalInterfaces`: stores the generic interval interface (`IInterval`), `ISerializedInterval`, `ISerializableInterval`
+
+- `Interval`: Implementation of the `ISerializableInterval`, whose endpoints are plain-old numbers
+
+- `SequenceInterval`: Implementation of the `ISerializableInterval`, whose ends are associated with positions in a mutatable sequence.
+
+3. Create a new folder for IntervalIndex-related components
+
+- `intervalIndexInterfaces`: stores the generic intervalIndex interface (`IIntervalIndex`), `IOverlappingIntervalsIndex`, `IStartpointInRangeIndex`, `IEndpointInRangeIndex`, and etc.
+
+- `StartpointInRangeIndex`
+
+- `EndpointInRangeIndex`
+
+- `OverlappingIntervalsIndex`
+
+- `OverlappingSequenceIntervalsIndex`
+
+- `SequenceIntervalIndexes`: store the namespace contains specialiazations of indexes which support spatial queries specifically for `SequenceInterval`s.
+
+4. Move `IntervalCollectionFactory` and `SequenceIntervalCollectionFactory` to separate files, which are responsible for creating instances of various types of IntervalCollection
+
+The files hierarchy is supposed to be:
+
+```
+├── IntervalCollection
+├── Interval
+│   ├── intervalInterfaces
+│   ├── Interval
+│   └── SequenceInterval
+├── IntervalIndex
+│   ├── intervalIndexInterfaces
+│   ├── startpointInRangeIndex
+│   ├── endpointInRangeIndex
+│   ├── overlappingIntervalsIndex
+│   ├── overlappingSequenceIntervalsIndex
+│   └── sequenceIntervalIndexes
+├── IntervalCollectionFactory
+├── SequenceIntervalCollectionFactory
+└── ... (other exisiting files)
+```
+
+## Deprecation
+
+1. Depreacate the unnecessary API
+
+- `SharedIntervalCollection` is not maintained and is planned to be removed
+
+- `IntervalConflictResolver` in `IntervalTree` and `addConflictResolver` in `intervalCollection` are never invoked and no longer needed, since interval collections support multiple intervals at the same location and gives each interval a unique id.
+
+- `propertyManager` of `Interval` and `SequenceInterval` is no longer needed
+
+2. Mark below API's as internal
+
+- `union`, `addProperties`, `modify` are never intended to be public
+
+3. The `idIntervalIndex` and `endIntervalIndex` were never used explicity, should we remove them from the `LocalIntervalCollection`?
```

Comment:
@Abe27342 May need confirmation on this point

Reply:
The idea is that we're going to remove the APIs that `LocalIntervalCollection` and `IntervalCollection` implement which use `endIntervalIndex` and `overlappingIntervalIndex`. Instead, we're going to export creation functions for those indexes and let the app author attach them themselves (if they want to do so). This is the bundle-size win the design focuses on.

Reply:
I anticipate `IIntervalCollection`'s API in next will go from this:

```typescript

/**
 * Collection of intervals that supports addition, modification, removal, and efficient spatial querying.
 * Changes to this collection will be incur updates on collaborating clients (i.e. they are not local-only).
 */
export interface IIntervalCollection<TInterval extends ISerializableInterval>
	extends TypedEventEmitter<IIntervalCollectionEvent<TInterval>> {
	readonly attached: boolean;
	/**
	 * Attaches an index to this collection.
	 * All intervals which are part of this collection will be added to the index, and the index will automatically
	 * be updated when this collection updates due to local or remote changes.
	 *
	 * @remarks - After attaching an index to an interval collection, applications should typically store this
	 * index somewhere in their in-memory data model for future reference and querying.
	 */
	attachIndex(index: IntervalIndex<TInterval>): void;
	/**
	 * Detaches an index from this collection.
	 * All intervals which are part of this collection will be removed from the index, and updates to this collection
	 * due to local or remote changes will no longer incur updates to the index.
	 *
	 * @returns - Return false if the target index cannot be found in the indexes, otherwise remove all intervals in the index and return true
	 */
	detachIndex(index: IntervalIndex<TInterval>): boolean;
	/**
	 * @returns the interval in this collection that has the provided `id`.
	 * If no interval in the collection has this `id`, returns `undefined`.
	 */
	getIntervalById(id: string): TInterval | undefined;
	/**
	 * Creates a new interval and add it to the collection.
	 * @param start - interval start position (inclusive)
	 * @param end - interval end position (exclusive)
	 * @param intervalType - type of the interval. All intervals are SlideOnRemove. Intervals may not be Transient.
	 * @param props - properties of the interval
	 * @param stickiness - {@link (IntervalStickiness:type)} to apply to the added interval.
	 * @returns - the created interval
	 * @remarks - See documentation on {@link SequenceInterval} for comments on interval endpoint semantics: there are subtleties
	 * with how the current half-open behavior is represented.
	 */
	add(
		start: number,
		end: number,
		intervalType: IntervalType,
		props?: PropertySet,
		stickiness?: IntervalStickiness,
	): TInterval;
	/**
	 * Removes an interval from the collection.
	 * @param id - Id of the interval to remove
	 * @returns the removed interval
	 */
	removeIntervalById(id: string): TInterval | undefined;
	/**
	 * Changes the properties on an existing interval.
	 * @param id - Id of the interval whose properties should be changed
	 * @param props - Property set to apply to the interval. Shallow merging is used between any existing properties
	 * and `prop`, i.e. the interval will end up with a property object equivalent to `{ ...oldProps, ...props }`.
	 */
	changeProperties(id: string, props: PropertySet);
	/**
	 * Changes the endpoints of an existing interval.
	 * @param id - Id of the interval to change
	 * @param start - New start value, if defined. `undefined` signifies this endpoint should be left unchanged.
	 * @param end - New end value, if defined. `undefined` signifies this endpoint should be left unchanged.
	 * @returns the interval that was changed, if it existed in the collection.
	 */
	change(id: string, start?: number, end?: number): TInterval | undefined;

	attachDeserializer(onDeserialize: DeserializeCallback): void;
	/**
	 * @returns an iterator over all intervals in this collection.
	 */
	[Symbol.iterator](): Iterator<TInterval>;

	/**
	 * @returns a forward iterator over all intervals in this collection with start point equal to `startPosition`.
	 */
	CreateForwardIteratorWithStartPosition(startPosition: number): Iterator<TInterval>;

	/**
	 * @returns a backward iterator over all intervals in this collection with start point equal to `startPosition`.
	 */
	CreateBackwardIteratorWithStartPosition(startPosition: number): Iterator<TInterval>;

	/**
	 * @returns a forward iterator over all intervals in this collection with end point equal to `endPosition`.
	 */
	CreateForwardIteratorWithEndPosition(endPosition: number): Iterator<TInterval>;

	/**
	 * @returns a backward iterator over all intervals in this collection with end point equal to `endPosition`.
	 */
	CreateBackwardIteratorWithEndPosition(endPosition: number): Iterator<TInterval>;

	/**
	 * Gathers iteration results that optionally match a start/end criteria into the provided array.
	 * @param results - Array to gather the results into. In lieu of a return value, this array will be populated with
	 * intervals matching the query upon edit.
	 * @param iteratesForward - whether or not iteration should be in the forward direction
	 * @param start - If provided, only match intervals whose start point is equal to `start`.
	 * @param end - If provided, only match intervals whose end point is equal to `end`.
	 */
	gatherIterationResults(
		results: TInterval[],
		iteratesForward: boolean,
		start?: number,
		end?: number,
	): void;

	/**
	 * @returns an array of all intervals in this collection that overlap with the interval
	 * `[startPosition, endPosition]`.
	 */
	findOverlappingIntervals(startPosition: number, endPosition: number): TInterval[];

	/**
	 * Applies a function to each interval in this collection.
	 */
	map(fn: (interval: TInterval) => void): void;

	previousInterval(pos: number): TInterval | undefined;

	nextInterval(pos: number): TInterval | undefined;
}

```

to this:

```typescript

/**
 * Collection of intervals that supports addition, modification, removal, and efficient spatial querying.
 * Changes to this collection will be incur updates on collaborating clients (i.e. they are not local-only).
 */
export interface IIntervalCollection<TInterval extends ISerializableInterval>
	extends TypedEventEmitter<IIntervalCollectionEvent<TInterval>> {
	readonly attached: boolean;
	/**
	 * Attaches an index to this collection.
	 * All intervals which are part of this collection will be added to the index, and the index will automatically
	 * be updated when this collection updates due to local or remote changes.
	 *
	 * @remarks - After attaching an index to an interval collection, applications should typically store this
	 * index somewhere in their in-memory data model for future reference and querying.
	 */
	attachIndex(index: IntervalIndex<TInterval>): void;
	/**
	 * Detaches an index from this collection.
	 * All intervals which are part of this collection will be removed from the index, and updates to this collection
	 * due to local or remote changes will no longer incur updates to the index.
	 *
	 * @returns - Return false if the target index cannot be found in the indexes, otherwise remove all intervals in the index and return true
	 */
	detachIndex(index: IntervalIndex<TInterval>): boolean;
	/**
	 * @returns the interval in this collection that has the provided `id`.
	 * If no interval in the collection has this `id`, returns `undefined`.
	 */
	getIntervalById(id: string): TInterval | undefined;
	/**
	 * Creates a new interval and add it to the collection.
	 * @param start - interval start position (inclusive)
	 * @param end - interval end position (exclusive)
	 * @param intervalType - type of the interval. All intervals are SlideOnRemove. Intervals may not be Transient.
	 * @param props - properties of the interval
	 * @param stickiness - {@link (IntervalStickiness:type)} to apply to the added interval.
	 * @returns - the created interval
	 * @remarks - See documentation on {@link SequenceInterval} for comments on interval endpoint semantics: there are subtleties
	 * with how the current half-open behavior is represented.
	 */
	add(
		start: number,
		end: number,
		intervalType: IntervalType,
		props?: PropertySet,
		stickiness?: IntervalStickiness,
	): TInterval;
	/**
	 * Removes an interval from the collection.
	 * @param id - Id of the interval to remove
	 * @returns the removed interval
	 */
	removeIntervalById(id: string): TInterval | undefined;
	/**
	 * Changes the properties on an existing interval.
	 * @param id - Id of the interval whose properties should be changed
	 * @param props - Property set to apply to the interval. Shallow merging is used between any existing properties
	 * and `prop`, i.e. the interval will end up with a property object equivalent to `{ ...oldProps, ...props }`.
	 */
	changeProperties(id: string, props: PropertySet);
	/**
	 * Changes the endpoints of an existing interval.
	 * @param id - Id of the interval to change
	 * @param start - New start value, if defined. `undefined` signifies this endpoint should be left unchanged.
	 * @param end - New end value, if defined. `undefined` signifies this endpoint should be left unchanged.
	 * @returns the interval that was changed, if it existed in the collection.
	 */
	change(id: string, start?: number, end?: number): TInterval | undefined;

	attachDeserializer(onDeserialize: DeserializeCallback): void;
	/**
	 * @returns an iterator over all intervals in this collection.
	 */
	[Symbol.iterator](): Iterator<TInterval>;

	/**
	 * Applies a function to each interval in this collection.
	 */
	map(fn: (interval: TInterval) => void): void;
}

```